Sources/Afluent/SequenceOperators/DeferredAsyncSequence.swift

```diff
@@ -0,0 +1,58 @@
+//
+//  DeferredAsyncSequence.swift
+//
+//
+//  Created by Annalise Mariottini on 12/20/23.
+//
+
+import Foundation
+
+/// An asynchronous sequence that defers the execution of another asynchronous sequence until iteration.
+///
+/// Using this type allows for an asynchronous sequence to be created without eagerly beginning execution.
+/// Notably, `AsyncStream` and `AsyncThrowingStream` immediately execute their passed closure to start yielding and buffering values.
+/// By wrapping either of these in a `DeferredSequence`, you can define the creation of an asynchronous sequence without executing it immediately.
+///
+/// Each time iteration begins using this type, the passed closure is called to create a new asynchronous sequence.
+/// This can allow for a sequence to be created and iterated over multiple times.
+///
+/// ```swift
+/// let deferred = DeferredSequence {
+///     AsyncStream { continuation in
+///         // yield some values asynchronously
+///     }
+/// }
+///
+/// for try await value in deferred {
+///     // starts at the first element
+/// }
+///
+/// for try await value in deferred {
+///     // starts at the first element
+/// }
+/// ```
+public struct DeferredAsyncSequence<Upstream: AsyncSequence>: AsyncSequence {
```

Comment:
Can you stick this on the AsyncSequences enum and remove "AsyncSequence" from the name?

We can make it friendly by creating a public type alias like `public typealias Deferred = AsyncSequences.Deferred` that should match other patterns in the repo